### DISCUSSION TITLE
Multiple GPU use? (Mixing Nvidia and AMD GPUs?) - Blender 4.0

### DISCUSSION THREAD

**Discussion by anthony0xtkpiq45ula** [vosen/ZLUDA#157] [Q&A]
Hello I have a quick question. Can ZLUDA be used in Blender 4.0 (or other programs) that utilize Cuda (or Optix) with multiple mixed NVIDIA and AMD GPUs?

I have an internal 4090FE and external 6900XT(connected by TB3).
Blender 4.0 can recognize, select and utilize between both GPUS without ZLUDA. (NVIDIA, OPTIX and CUDA okay) (AMD, HIP okay)
Using ZLUDA Blender 4.0 only recognizes and utilizes the AMD GPU with CUDA or OPTIXs.

If it is possible, what args or options do I append to the commandline to utilize both GPUS with CUDA or OPTIXS together in blender 4.0 (or other programs).

Thank you for your time.

**Comment by lshqqytiger** [comment]
In order to achieve that, we should be able to load multiple duplicated dll files like nvcuda.dll. But it is hard unless we have source code to build entire Blender. In another way, if ZLUDA can detect both NVIDIA/AMD gpus and connect its api to corresponding CUDA/HIP apis, we can utilize them simultaneously. However, it is very difficult to make both gpus work seamlessly. I don't think this is possible unless there are layers that abstract CUDA and HIP.

**Comment by vosen** [comment]
This should be possible. As long as the original CUDA driver exists we can load it (it gets tricky on Windows, because we would have to pick the right nvcuda64.dll from the drivers directories) explicitly at runtime with `LoadLibary`/`dlopen`. Duplicate file names are not a problem (in fact, right now if you are using zluda_dump you will have duplicate nvcuda.dlls in your process). It's still more work than just loading the dll, API functions need to be dispatched correctly and I presume a new joint ZLUDA-CUDA context type that "sees" both kind of GPUs need to be written. How seamless this works is still under question. I am not sure if multiple NV GPUs are expected to be able to do anything special (sharing memory?).

ZLUDA might get this feature in the future. It's just that it is a rare setup. The common NVIDIA/AMD setup is dedicated NV card + integrated AMD card, in which case you probably don't need ZLUDA.

**Comment by NikolaB131** [comment]
I have a GTX1080 and an old RX560, it would be cool if I could choose both of them with CUDA.

**Comment by tarunspandit** [comment]
Any updates on this?
